Framework/Algorithms/test/ExtractSingleSpectrumTest.h

```diff
@@ -109,17 +109,13 @@ class ExtractSingleSpectrumTest : public CxxTest::TestSuite {
   void do_Spectrum_Tests(MatrixWorkspace_sptr outputWS, const specnum_t specID,
                          const detid_t detID) {
     TS_ASSERT_EQUALS(outputWS->getNumberHistograms(), 1);
-    const Mantid::API::ISpectrum *spectrum(NULL);
-    TS_ASSERT_THROWS_NOTHING(spectrum = outputWS->getSpectrum(0));
-    if (spectrum) {
-      TS_ASSERT_EQUALS(spectrum->getSpectrumNo(), specID);
-      auto detids = spectrum->getDetectorIDs();
-      TS_ASSERT_EQUALS(detids.size(), 1);
-      const detid_t id = *(detids.begin());
-      TS_ASSERT_EQUALS(id, detID);
-    } else {
-      TS_FAIL("No spectra/detectors associated with extracted histogram.");
-    }
+    TS_ASSERT_THROWS_NOTHING(outputWS->getSpectrum(0));
+    const auto &spectrum = outputWS->getSpectrum(0);
```

Comment:
@KarlPalmen Good question.

If you look at the existing workspace code, there seems to be no way of creating a workspace with no spectra (constructors throw, etc.), i.e., one spectrum is the minimum and thus `getSpectrum(0)` should never fail by definition.

Furthermore, the implementations of `getSpectrum` would actually never return `nullptr`. Instead, if they do a range check they throw (e.g., for `Workspace2D`).

So in summary: Previously, `getSpectrum` mixed two concepts of error handling: exceptions and return of a `NULL` object (pointer). Only the former was actually used, the checks for `NULL` in some part of the code were either artefacts from an earlier implementation without exceptions, or they were added because the previous interface was confusing for developers (typically, if a function returns a pointer we assume that it might be `NULL`). Be switching to return by reference, we remove this confusion. Indices that are out of range should be handled by throwing exceptions.